{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import urllib.request\n",
    "import gzip\n",
    "import shutil\n",
    "import os\n",
    "import pickle\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "from numpy.random import choice\n",
    "import pandas as pd\n",
    "from hmmlearn.hmm import MultinomialHMM\n",
    "from scipy.stats import ttest_rel\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### An HMM to model codon usage in Bacillus subtilis.\n",
    "\n",
    "In this notebook we go through the process of building a hidden markov model (HMM) to approximate the codon use distribution of *Bacillus subtilis*. We treat codons as the hidden states (because those are the things we want to predict) and amino acids as the observed features. One practical use of a model like this would be to optimize a heterologous (from another host) DNA sequence for expression in *B. subtilis*. \n",
    "\n",
    "First, we download the *B. subtilis genome*. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "url = 'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/009/045/GCF_000009045.1_ASM904v1/GCF_000009045.1_ASM904v1_genomic.gbff.gz'\n",
    "gz_fn = 'GCF_000009045.1_ASM904v1_genomic.gbff.gz'\n",
    "gbk_fn = gz_fn.replace('.gbff.gz', '.gbk')\n",
    "if not os.path.isfile(gz_fn):\n",
    "    print('Beginning file download with urllib2...')\n",
    "    urllib.request.urlretrieve(url, gz_fn)\n",
    "\n",
    "if not os.path.isfile(gbk_fn):\n",
    "    with gzip.open(gz_fn, 'rb') as f_in, open(gbk_fn, 'wb') as f_out:\n",
    "            shutil.copyfileobj(f_in, f_out)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Below are some helper functions to split a nucleotide sequence into codons and check for ambiguous codons or sequences with lengths that aren't a multiple of 3.\n",
    "\n",
    "`assert` statements are little tests to make sure that things are working the way we expect. They are *very* helpful for catching silly bugs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_list_of_codons(dna_seq):\n",
    "    codons = []\n",
    "    for i in range(0, len(dna_seq), 3):\n",
    "        codons.append(dna_seq[i:i+3])\n",
    "    return codons\n",
    "assert get_list_of_codons('ATGCCCGGGAAATTTTAG') == ['ATG', 'CCC', 'GGG', 'AAA', 'TTT', 'TAG']\n",
    "\n",
    "def check_len_and_ambiguity(seq):\n",
    "    assert isinstance(seq, str)\n",
    "    ambig_nucs = ['R', 'Y', 'S', 'W', 'K', 'M', 'B', 'D', 'H', 'V', 'N']\n",
    "    unambiguous = not any([anuc in seq for anuc in ambig_nucs])\n",
    "    multiple_3 = len(seq) % 3 == 0\n",
    "    return unambiguous and multiple_3\n",
    "assert check_len_and_ambiguity('ATGACCTAG')\n",
    "assert not check_len_and_ambiguity('ATGACCTA')\n",
    "assert not check_len_and_ambiguity('ATGACCTAY')\n",
    "assert not check_len_and_ambiguity('ATGACCAY')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And now we read in all the protein coding sequences from the genome. During this process, we keep track of the set of unique amino acids and nucleotides, the initial states, and---of great importance---the codons that correspond to each amino acid."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "translations = {}\n",
    "for record in SeqIO.parse(gbk_fn, \"genbank\"):\n",
    "    for feature in record.features:\n",
    "        if feature.type == 'CDS' and \\\n",
    "           'translation' in feature.qualifiers and \\\n",
    "           check_len_and_ambiguity(str(record.seq)):\n",
    "            protein = feature.qualifiers['translation'][0] + '*'\n",
    "            aas = {aa for aa in protein}\n",
    "            nucleotide_seq = str(feature.extract(record.seq))\n",
    "            if len(protein)*3 == len(nucleotide_seq):\n",
    "                translations[protein] = nucleotide_seq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, for purposes of evaluating our model later, we randomly split our set of coding sequences in training and test sets."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "TEST_FRACTION = 0.2\n",
    "NUM_SEQS = len(translations)\n",
    "np.random.seed(2021)\n",
    "\n",
    "# Split protein/DNA pairs into random training and test set\n",
    "test_indices = choice(NUM_SEQS, \n",
    "                      size=int(np.floor(TEST_FRACTION*NUM_SEQS)), \n",
    "                      replace=False)\n",
    "train_indices = [x for x in range(NUM_SEQS) if x not in test_indices]\n",
    "trnslts_lst = list(translations.items())\n",
    "test_translations = dict([trnslts_lst[x] for x in test_indices])\n",
    "train_translations = dict([trnslts_lst[x] for x in train_indices])\n",
    "\n",
    "# Save the training and test data to\n",
    "# train and test the RNN in notebook 2\n",
    "with open('test_translations.pickle', 'wb') as handle:\n",
    "    pickle.dump(test_translations, handle)\n",
    "    \n",
    "with open('train_translations.pickle', 'wb') as handle:\n",
    "    pickle.dump(train_translations, handle)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here we go about the business of counting features and states, with the end goal of building an HMM model."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "u_aas = set()    # set of \"unique\" amino acids\n",
    "u_codons = set() # set of \"unique\" codons\n",
    "all_codons = []\n",
    "initial_states = []\n",
    "emissions = {}\n",
    "for protein in train_translations.keys():\n",
    "    nucleotide_seq = train_translations[protein]\n",
    "    aas = {aa for aa in protein}\n",
    "    codon = get_list_of_codons(nucleotide_seq)\n",
    "    if len(protein) == len(codon):\n",
    "        all_codons.append(codon)\n",
    "        initial_states.append(codon[0])\n",
    "        u_aas = u_aas.union(aas)\n",
    "        u_codons = u_codons.union(set(codon))\n",
    "        for i, cdn in enumerate(codon):\n",
    "            emissions[cdn] = protein[i]\n",
    "lu_aas = list(u_aas)\n",
    "lu_codons = list(u_codons)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's look at some of the objects we created."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['AAC', 'AGT', 'GAG']"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lu_codons[:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('K', 'M')"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "emissions['AAA'], emissions['ATG']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Helper functions to encode and decode sequences (remember, our HMM works with integers)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def encode_seq(seq_obj, seqtype='dna'):\n",
    "    encdr = lu_codons\n",
    "    symbols = get_list_of_codons(seq_obj)\n",
    "    if seqtype != 'dna':\n",
    "        encdr = lu_aas\n",
    "        symbols = [c for c in seq_obj]\n",
    "    outseq = np.array([encdr.index(s) for s in symbols])\n",
    "    return outseq\n",
    "\n",
    "test_aa = 'MENILD'\n",
    "test_nuc = 'AAAAAAATAAGATAG'\n",
    "assert encode_seq(test_aa, seqtype='prot')[0] == lu_aas.index(test_aa[0]) and \\\n",
    "       encode_seq(test_aa, seqtype='prot')[-1] == lu_aas.index(test_aa[-1])\n",
    "assert encode_seq(test_nuc, seqtype='dna')[0] == lu_codons.index(test_nuc[0:3]) and \\\n",
    "       encode_seq(test_nuc, seqtype='dna')[-1] == lu_codons.index(test_nuc[-3:])\n",
    "\n",
    "def decode_seq(num_array, seqtype='dna'):\n",
    "    encdr = lu_codons\n",
    "    if seqtype != 'dna':\n",
    "        encdr = lu_aas\n",
    "    outseq = [encdr[s] for s in num_array]\n",
    "    return ''.join(outseq)\n",
    "\n",
    "assert decode_seq(encode_seq(test_nuc)) == test_nuc\n",
    "assert decode_seq(encode_seq(test_aa, seqtype='prot'), seqtype='prot') == test_aa"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we construct the three matrices that together form our HMM. \n",
    "\n",
    "The `emission_prob` matrix maps the likelihood of a particular codon emitting a particular amino acid. This is pretty simple, since a given codon will map to its amino acid 100% of the time."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "emission_prob = np.zeros((len(lu_codons), len(lu_aas)))\n",
    "for i, codon in enumerate(lu_codons):\n",
    "    aa = emissions[codon]\n",
    "    j = lu_aas.index(aa)\n",
    "    emission_prob[i, j] = 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The `initial_probabilities` matrix counts the number of times each codon is found at the start of a sequence, then divides those by the total so that they sum to 1. We can verify that \"ATG\" is a high fraction of the start probability."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "initial_probabilities = {k:0 for k in lu_codons}\n",
    "for i in initial_states:\n",
    "    initial_probabilities[i] += 1\n",
    "for k in initial_probabilities.keys():\n",
    "    initial_probabilities[k] = initial_probabilities[k] / len(initial_states)\n",
    "initial_probs_np = np.array([initial_probabilities[x] for x in lu_codons])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.7701386839775745"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "atg_idx = lu_codons.index('ATG')\n",
    "initial_probs_np[atg_idx]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finally, we compute the transition likelihoods. This is where the HMM differs substantially from a simpler approach of using only the most common codons. Here the HMM takes into account the codons that are expected immediately before or after."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AAC</th>\n",
       "      <th>AGT</th>\n",
       "      <th>GAG</th>\n",
       "      <th>GTG</th>\n",
       "      <th>GGG</th>\n",
       "      <th>TTG</th>\n",
       "      <th>CAC</th>\n",
       "      <th>GTT</th>\n",
       "      <th>CGG</th>\n",
       "      <th>ATC</th>\n",
       "      <th>...</th>\n",
       "      <th>CTG</th>\n",
       "      <th>CGC</th>\n",
       "      <th>TGT</th>\n",
       "      <th>TGA</th>\n",
       "      <th>ATG</th>\n",
       "      <th>GAC</th>\n",
       "      <th>ATT</th>\n",
       "      <th>CAT</th>\n",
       "      <th>ACG</th>\n",
       "      <th>TCG</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAC</th>\n",
       "      <td>0.020839</td>\n",
       "      <td>0.011675</td>\n",
       "      <td>0.014710</td>\n",
       "      <td>0.013251</td>\n",
       "      <td>0.017629</td>\n",
       "      <td>0.009398</td>\n",
       "      <td>0.007413</td>\n",
       "      <td>0.013776</td>\n",
       "      <td>0.010099</td>\n",
       "      <td>0.028370</td>\n",
       "      <td>...</td>\n",
       "      <td>0.022065</td>\n",
       "      <td>0.016520</td>\n",
       "      <td>0.004086</td>\n",
       "      <td>0.000817</td>\n",
       "      <td>0.024634</td>\n",
       "      <td>0.013192</td>\n",
       "      <td>0.036017</td>\n",
       "      <td>0.014010</td>\n",
       "      <td>0.012550</td>\n",
       "      <td>0.003911</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGT</th>\n",
       "      <td>0.006827</td>\n",
       "      <td>0.001517</td>\n",
       "      <td>0.037172</td>\n",
       "      <td>0.032317</td>\n",
       "      <td>0.002731</td>\n",
       "      <td>0.019572</td>\n",
       "      <td>0.003793</td>\n",
       "      <td>0.027765</td>\n",
       "      <td>0.002124</td>\n",
       "      <td>0.022758</td>\n",
       "      <td>...</td>\n",
       "      <td>0.020331</td>\n",
       "      <td>0.001669</td>\n",
       "      <td>0.002276</td>\n",
       "      <td>0.000607</td>\n",
       "      <td>0.026248</td>\n",
       "      <td>0.032469</td>\n",
       "      <td>0.042330</td>\n",
       "      <td>0.014717</td>\n",
       "      <td>0.011227</td>\n",
       "      <td>0.007434</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GAG</th>\n",
       "      <td>0.012388</td>\n",
       "      <td>0.004505</td>\n",
       "      <td>0.024386</td>\n",
       "      <td>0.016719</td>\n",
       "      <td>0.006757</td>\n",
       "      <td>0.006540</td>\n",
       "      <td>0.011132</td>\n",
       "      <td>0.020098</td>\n",
       "      <td>0.014987</td>\n",
       "      <td>0.018885</td>\n",
       "      <td>...</td>\n",
       "      <td>0.057565</td>\n",
       "      <td>0.013601</td>\n",
       "      <td>0.001992</td>\n",
       "      <td>0.000823</td>\n",
       "      <td>0.020141</td>\n",
       "      <td>0.017239</td>\n",
       "      <td>0.030797</td>\n",
       "      <td>0.030277</td>\n",
       "      <td>0.012908</td>\n",
       "      <td>0.004331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTG</th>\n",
       "      <td>0.018153</td>\n",
       "      <td>0.004920</td>\n",
       "      <td>0.016174</td>\n",
       "      <td>0.010801</td>\n",
       "      <td>0.005090</td>\n",
       "      <td>0.009501</td>\n",
       "      <td>0.006843</td>\n",
       "      <td>0.010519</td>\n",
       "      <td>0.006390</td>\n",
       "      <td>0.032121</td>\n",
       "      <td>...</td>\n",
       "      <td>0.056665</td>\n",
       "      <td>0.007352</td>\n",
       "      <td>0.002488</td>\n",
       "      <td>0.000735</td>\n",
       "      <td>0.029916</td>\n",
       "      <td>0.011876</td>\n",
       "      <td>0.050161</td>\n",
       "      <td>0.011423</td>\n",
       "      <td>0.020585</td>\n",
       "      <td>0.005655</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GGG</th>\n",
       "      <td>0.016482</td>\n",
       "      <td>0.005912</td>\n",
       "      <td>0.015944</td>\n",
       "      <td>0.012092</td>\n",
       "      <td>0.005106</td>\n",
       "      <td>0.008420</td>\n",
       "      <td>0.009674</td>\n",
       "      <td>0.019437</td>\n",
       "      <td>0.007435</td>\n",
       "      <td>0.025170</td>\n",
       "      <td>...</td>\n",
       "      <td>0.050072</td>\n",
       "      <td>0.010301</td>\n",
       "      <td>0.002418</td>\n",
       "      <td>0.001254</td>\n",
       "      <td>0.028843</td>\n",
       "      <td>0.012092</td>\n",
       "      <td>0.054640</td>\n",
       "      <td>0.022752</td>\n",
       "      <td>0.014690</td>\n",
       "      <td>0.004210</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 64 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          AAC       AGT       GAG       GTG       GGG       TTG       CAC  \\\n",
       "AAC  0.020839  0.011675  0.014710  0.013251  0.017629  0.009398  0.007413   \n",
       "AGT  0.006827  0.001517  0.037172  0.032317  0.002731  0.019572  0.003793   \n",
       "GAG  0.012388  0.004505  0.024386  0.016719  0.006757  0.006540  0.011132   \n",
       "GTG  0.018153  0.004920  0.016174  0.010801  0.005090  0.009501  0.006843   \n",
       "GGG  0.016482  0.005912  0.015944  0.012092  0.005106  0.008420  0.009674   \n",
       "\n",
       "          GTT       CGG       ATC  ...       CTG       CGC       TGT  \\\n",
       "AAC  0.013776  0.010099  0.028370  ...  0.022065  0.016520  0.004086   \n",
       "AGT  0.027765  0.002124  0.022758  ...  0.020331  0.001669  0.002276   \n",
       "GAG  0.020098  0.014987  0.018885  ...  0.057565  0.013601  0.001992   \n",
       "GTG  0.010519  0.006390  0.032121  ...  0.056665  0.007352  0.002488   \n",
       "GGG  0.019437  0.007435  0.025170  ...  0.050072  0.010301  0.002418   \n",
       "\n",
       "          TGA       ATG       GAC       ATT       CAT       ACG       TCG  \n",
       "AAC  0.000817  0.024634  0.013192  0.036017  0.014010  0.012550  0.003911  \n",
       "AGT  0.000607  0.026248  0.032469  0.042330  0.014717  0.011227  0.007434  \n",
       "GAG  0.000823  0.020141  0.017239  0.030797  0.030277  0.012908  0.004331  \n",
       "GTG  0.000735  0.029916  0.011876  0.050161  0.011423  0.020585  0.005655  \n",
       "GGG  0.001254  0.028843  0.012092  0.054640  0.022752  0.014690  0.004210  \n",
       "\n",
       "[5 rows x 64 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transition_counts = np.ones((len(lu_codons), len(lu_codons)))\n",
    "for gene in all_codons:\n",
    "    for i in range(0, len(gene)-1):\n",
    "        codon0 = lu_codons.index(gene[i])\n",
    "        codon1 = lu_codons.index(gene[i+1])\n",
    "        transition_counts[codon0, codon1] += 1\n",
    "transition_totals = transition_counts.sum(axis=1)\n",
    "transition_probs = np.dot(np.diag(1/transition_totals), transition_counts)\n",
    "transition_probs_df = pd.DataFrame(transition_probs, index=lu_codons, columns=lu_codons)\n",
    "transition_probs_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Build the HMM\n",
    "\n",
    "In this case, rather than train the HMM (which is [also an option using the `model.fit()` function](https://hmmlearn.readthedocs.io/en/latest/tutorial.html#training-hmm-parameters-and-inferring-the-hidden-states)), we plug in our own calculated matrices to create the model object."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "hmm = MultinomialHMM(n_components=len(lu_codons), \n",
    "                     startprob_prior=initial_probs_np, \n",
    "                     transmat_prior=transition_probs, \n",
    "                     verbose=False,  \n",
    "                     init_params='')\n",
    "hmm.transmat_ = transition_probs\n",
    "hmm.emissionprob_ = emission_prob\n",
    "hmm.startprob_ = initial_probs_np\n",
    "hmm.n_features = len(lu_aas)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Having built the model, we can sample a randomly-generated sequence and check that everything makes sense."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_aa, sample_nuc = hmm.sample(n_samples=10, random_state=21)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('VASSPHIFVK', 'GTGGCGTCATCTCCGCACATCTTTGTAAAA')"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "assert emission_prob[sample_nuc[0], sample_aa[0]] == 1\n",
    "decode_seq(sample_aa.reshape(-1), seqtype='prot'), decode_seq(sample_nuc.reshape(-1), seqtype='dna')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### HMM performance\n",
    "\n",
    "Here we extract a single protein, predict the most likely codon usage under our HMM, and see how well it compares to the native DNA sequence."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'MKFLLGNINSTVLTMAGLRVLSSMIELTAAIVMLVTNDVRKAVVVNSILAIVGPLIFIITMTVGIYQIAGQLSYAKLILIFTGVVLILAGVHK*'"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_protein = list(test_translations.keys())[10]\n",
    "test_dna = test_translations[test_protein]\n",
    "test_protein"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "prot_e = encode_seq(test_protein, seqtype='prot')\n",
    "mle_dna_indices = hmm.predict(prot_e.reshape(-1, 1))\n",
    "pred_dna = decode_seq(mle_dna_indices)\n",
    "\n",
    "assert str(Seq(pred_dna).translate()) == test_protein"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Not only can we predict the most likely codon at each amino acid position, but we can also predict the probability distribution among all the codons. For example, at index location 4 in the current protein, we can see the probability distribution:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Codon\tP(Codon)\n",
      "TTG\t0.10\n",
      "CTC\t0.25\n",
      "CTT\t0.26\n",
      "TTA\t0.20\n",
      "CTA\t0.05\n",
      "CTG\t0.13\n"
     ]
    }
   ],
   "source": [
    "state_probabilities = hmm.predict_proba(prot_e.reshape(-1, 1))\n",
    "print('Codon\\tP(Codon)')\n",
    "for i, prob in enumerate(state_probabilities[4,]):\n",
    "    codon = lu_codons[i]\n",
    "    if prob > 0:\n",
    "        print(f'{codon}\\t{prob:.2f}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can see that \"CTA\" is the least likely in this context, and \"CTT\" is the most likely (and is the one we see in both the predicted and native sequences)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Seq1 ATGAAATTTTTACTTGGAAATATCAATTCTACTGTTTTAACAATGGCGGGATTACGAGTT\n",
      "                *              ****  *  ** *           ** *  *  *\n",
      "Seq2 ATGAAATTTTTGCTTGGAAATATCAACAGCACAGTGCTGACAATGGCGGGGCTTCGTGTG\n",
      "\n",
      "Seq1 TTATCCTCTATGATTGAGCTGACGGCAGCAATTGTCATGCTTGTGACCAACGATGTCCGG\n",
      "     * *  *  *              *  *  *                 *  *      * *\n",
      "Seq2 CTTTCTTCAATGATTGAGCTGACAGCTGCCATTGTCATGCTTGTGACAAATGATGTCAGA\n",
      "\n",
      "Seq1 AAGGCGGTTGTGGTTAACAGCATTCTCGCTATTGTCGGTCCGTTGATTTTTATCATTACA\n",
      "       *        *  *           *  *        *   *                *\n",
      "Seq2 AAAGCGGTTGTTGTCAACAGCATTCTTGCCATTGTCGGCCCGCTGATTTTTATCATTACG\n",
      "\n",
      "Seq1 ATGACTGTCGGAATCTACCAAATTGCCGGGCAGCTTTCGTATGCAAAGCTGATTCTGATT\n",
      "          *     *  *  *           *        *               *     \n",
      "Seq2 ATGACAGTCGGCATTTATCAAATTGCCGGACAGCTTTCTTATGCAAAGCTGATTTTGATT\n",
      "\n",
      "Seq1 TTTACGGGTGTTGTCTTGATTTTGGCGGGTGTTCATAAATAA\n",
      "          *  *     **     * *  *               \n",
      "Seq2 TTTACAGGCGTTGTGCTGATTCTTGCCGGTGTTCATAAATAA\n",
      "\n"
     ]
    }
   ],
   "source": [
    "def align(seq1, seq2, WIDTH=60):\n",
    "    '''Align two input sequences of equal length,\n",
    "    with *  between indicating mismatches.'''\n",
    "    lines = int(np.ceil(len(seq1) / WIDTH))\n",
    "    match = ''\n",
    "    for i, c1 in enumerate(seq1):\n",
    "        indicator = ' '\n",
    "        if c1 != seq2[i]:\n",
    "            indicator = '*'\n",
    "        match += indicator\n",
    "    \n",
    "    for i in range(lines):\n",
    "        print('Seq1', seq1[i*WIDTH:i*WIDTH+WIDTH])\n",
    "        print('    ', match[i*WIDTH:i*WIDTH+WIDTH])\n",
    "        print('Seq2', seq2[i*WIDTH:i*WIDTH+WIDTH])\n",
    "        print()\n",
    "\n",
    "align(test_dna, pred_dna)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "That's a nice visual, to see so many nucleotides matching! But how does it look quantitatively?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HMM fraction bps matching: 0.833\n"
     ]
    }
   ],
   "source": [
    "def fraction_matches(seq1, seq2):\n",
    "    '''Compare sequences in terms of matching nucleotides.'''\n",
    "    matches = 0\n",
    "    for i, c1 in enumerate(seq1):\n",
    "        if c1 == seq2[i]:\n",
    "            matches += 1\n",
    "    return {'length':len(seq1), \n",
    "            'n_matches':matches, \n",
    "            'n_mismatches':len(seq1) - matches, \n",
    "            'fraction_matches':matches/len(seq1)}\n",
    "\n",
    "fm_hmm = fraction_matches(test_dna, pred_dna)\n",
    "\n",
    "print(f'HMM fraction bps matching: {fm_hmm[\"fraction_matches\"]:.3f}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So 83% of predicted nucleotides match those actually found to code for this protein. Is that good model performance?\n",
    "\n",
    "One comparison would be to compare our predictions with random DNA sequences, but that's not very interesting. A better comparison is to benchmark with a standard approach to codon optimization: which is, to always predict the most common codon for any given amino acid.\n",
    "\n",
    "For that, we'll want a codon usage table."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>amino_acid</th>\n",
       "      <th>codon</th>\n",
       "      <th>frequency</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>V</td>\n",
       "      <td>GTG</td>\n",
       "      <td>0.257938</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>V</td>\n",
       "      <td>GTT</td>\n",
       "      <td>0.287079</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>W</td>\n",
       "      <td>TGG</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>Y</td>\n",
       "      <td>TAC</td>\n",
       "      <td>0.347468</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>Y</td>\n",
       "      <td>TAT</td>\n",
       "      <td>0.652532</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   amino_acid codon  frequency\n",
       "64          V   GTG   0.257938\n",
       "65          V   GTT   0.287079\n",
       "66          W   TGG   1.000000\n",
       "67          Y   TAC   0.347468\n",
       "68          Y   TAT   0.652532"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def build_codon_usage_table(prot_dna_dict):\n",
    "    '''Returns a codon usage table.'''\n",
    "    cu_dict = {'amino_acid':[],\n",
    "               'codon':[]}\n",
    "    for prot in prot_dna_dict.keys():\n",
    "        codons = get_list_of_codons(prot_dna_dict[prot])\n",
    "        for i, aa in enumerate(prot):\n",
    "            cu_dict['amino_acid'].append(aa)\n",
    "            cu_dict['codon'].append(codons[i])\n",
    "    cu_df = pd.DataFrame(cu_dict)\n",
    "    cu_df = (cu_df\n",
    "             .groupby(['amino_acid', 'codon'])\n",
    "             .size()\n",
    "             .reset_index()\n",
    "            )\n",
    "    cu_df['sum_aa'] = (cu_df\n",
    "                       .groupby('amino_acid')[0]\n",
    "                       .transform('sum')\n",
    "                      )\n",
    "    cu_df = (cu_df\n",
    "             .assign(frequency = lambda x: x[0] / x['sum_aa'])\n",
    "            )\n",
    "    \n",
    "    return cu_df[['amino_acid', 'codon', 'frequency']]\n",
    "    \n",
    "cu_table = build_codon_usage_table(train_translations)\n",
    "cu_table.to_csv('train_codon_usage.tsv', sep='\\t', index=False)\n",
    "cu_table.tail(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_most_common_codons(protein, codon_usage_table):\n",
    "    '''Return a DNA sequence using the most frequent codons.'''\n",
    "    mlaa = (codon_usage_table\n",
    "            .sort_values(by='frequency', ascending=False)\n",
    "            .drop_duplicates(subset=['amino_acid'])\n",
    "           )\n",
    "    dna = ''\n",
    "    for aa in protein:\n",
    "        dna += mlaa.query(f'amino_acid == \"{aa}\"').codon.values[0]\n",
    "    return dna\n",
    "\n",
    "assert get_most_common_codons('YWY', cu_table) == 'TATTGGTAT'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, how does the codon usage model compare to the HMM on this test protein?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Codon usage fraction bps matching: 0.780\n"
     ]
    }
   ],
   "source": [
    "cu_dna = get_most_common_codons(test_protein, cu_table)\n",
    "fm_cu = fraction_matches(test_dna, cu_dna)\n",
    "\n",
    "print(f'Codon usage fraction bps matching: {fm_cu[\"fraction_matches\"]:.3f}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For this example, the HMM outperformed the naive approach of just assigning the most common codons!\n",
    "\n",
    "What about for a larger test set?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "TEST_N = min(100, len(test_translations))\n",
    "hmm_fms = np.zeros(TEST_N)\n",
    "cu_fms = np.zeros(TEST_N)\n",
    "for i, test_protein in enumerate(test_translations.keys()):\n",
    "    if i >= TEST_N:\n",
    "        break\n",
    "        \n",
    "    test_dna = test_translations[test_protein]\n",
    "    \n",
    "    # HMM\n",
    "    prot_e = encode_seq(test_protein, seqtype='prot')\n",
    "    mle_dna_indices = hmm.predict(prot_e.reshape(-1, 1))\n",
    "    pred_dna = decode_seq(mle_dna_indices)\n",
    "    fm_hmm = fraction_matches(test_dna, pred_dna)\n",
    "    hmm_fms[i] = fm_hmm[\"fraction_matches\"]\n",
    "    \n",
    "    # Codon usage\n",
    "    cu_dna = get_most_common_codons(test_protein, cu_table)\n",
    "    fm_cu = fraction_matches(test_dna, cu_dna)\n",
    "    cu_fms[i] = fm_cu[\"fraction_matches\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HMM: 0.806, CU: 0.794\n"
     ]
    }
   ],
   "source": [
    "print(f'HMM: {np.mean(hmm_fms):.3f}, CU: {np.mean(cu_fms):.3f}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAX4AAAEGCAYAAABiq/5QAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8QVMy6AAAACXBIWXMAAAsTAAALEwEAmpwYAAAZmklEQVR4nO3dfZRcdZ3n8ffH0NBBkpCHlg0msaM8CCo20uADszMIzIis0cQTCazDBtbZyCw4RkYHjnrYxr/AwYFd2cENyCQCkxAwPIRBhGGSiAqYDtNpgsGBCUEaIokBNIwGSfjuH/fXodLp6lR3162q7vt5nVOnb/3qPnyquu+3bv/q1u8qIjAzs+J4S70DmJlZbbnwm5kVjAu/mVnBuPCbmRWMC7+ZWcEcUO8AlZgyZUq0trbWO4aZ2Yiybt26X0dES9/2EVH4W1tb6ezsrHcMM7MRRdKz/bW7q8fMrGBc+M3MCsaF38ysYEZEH7+Z2WC8/vrr9PT0sHPnznpHqYnm5mamTZtGU1NTRfO78JvZqNPT08O4ceNobW1FUr3j5Coi2L59Oz09PcycObOiZdzVY2ajzs6dO5k8efKoL/oAkpg8efKg/rtx4TezUakIRb/XYJ+rC7+ZWcG4j9/MRr2Ojtqv75BDDuHVV1/dc3/x4sV0dnZy7bXX0tHRweWXX85TTz3FEUccAcDVV1/NxRdfzNq1a2lvb6e1tZXp06fz0EMP7VlHW1sbu3btYsOGDcPK78Jv1dPdUb9tH1fHbZsNwfve9z6WLVvG17/+dQBuv/12jj322L3m2bFjB8899xzTp09n48aNVdu2u3rMzOpg9uzZ3HXXXQBs2rSJCRMm0NKy97A6Z511FrfeeisAS5cu5ZxzzqnKtl34zcxy8Pvf/562trY9t8suu2yvx8ePH8/06dPZsGEDS5cuZd68efusY+7cuaxYsQKAlStXMmvWrKpkc+E3M8vB2LFj6erq2nP7xje+sc88Z599NsuWLePOO+9kzpw5+zw+adIkJk6cyLJlyzjmmGM4+OCDq5LNhd/MrE5mzZrFTTfdxIwZMxg/fny/88ybN48LL7ywat084A93zczqZuzYsVx55ZUcddRRZeeZM2cOW7Zs4WMf+xgvvPBCVbbrwm9mo161T+esprPPPnvAx8eNG8cll1xS1W0qIqq6wjy0t7eHL8QyAnR31G/bPp3TSmzcuJFjjjmm3jFqqr/nLGldRLT3nTe3Pn5JzZJ+Jmm9pCckXZ7aOyQ9L6kr3c7MK4OZme0rz66e14BTI+JVSU3AjyX9ID12dURcleO2zcysjNwKf2R9SL3fV25Kt8bvVzIzG+VyPZ1T0hhJXcBW4IGIeDQ9dJGkbkk3SppYZtkFkjoldW7bti3PmGZmhZJr4Y+I3RHRBkwDTpL0XuA64F1AG7AF+FaZZRdFRHtEtPf9GrOZmQ1dTb7AFRGvAKuBMyLixfSG8AZwPXBSLTKYmVkmtz5+SS3A6xHxiqSxwOnAlZKmRsSWNNscYHjji5qZ7U93R3XXV8Hpw7/61a9YuHAha9eu5aCDDqK1tZXZs2dz9913c8899+yZ77zzzuMTn/gEc+fOrW7GAeR5Vs9UYImkMWT/WSyPiHsk3SSpjeyD3s3A53PMYGZWcxHBnDlzmD9/PsuWLQOgq6uLlStX1jlZJs+zerqB4/tpPzevbZqZNYJVq1bR1NTEBRdcsKetra2NV155hUcffXSAJWvDg7SZmVXZhg0bOOGEE+odoywXfjOzGil3UfRaXxjehd/MrMre8573sG7dun3aJ0+ezMsvv7xX20svvcSUKVNqFQ1w4Tczq7pTTz2V1157jeuvv35P29q1a9m+fTsvvPDCnuvnPvvss6xfv562traa5vOwzGY2+tV49FZJ3HHHHSxcuJArrriC5uZmWltbueaaa7j55ps5//zz2blzJ01NTdxwww1MmDChpvlc+M3McnD44YezfPnyfdqPPPJIHnnkkTokepO7eszMCsaF38ysYFz4zWxUGglXF6yWwT5XF34zG3Wam5vZvn17IYp/RLB9+3aam5srXsYf7prZqDNt2jR6enooyrU8mpubmTZtWsXzu/Cb2ajT1NTEzJkz6x2jYbmrx8ysYFz4zcwKxoXfzKxgXPjNzArGhd/MrGBc+M3MCsaF38ysYHIr/JKaJf1M0npJT0i6PLVPkvSApKfSz4l5ZTAzs33lecT/GnBqRLwfaAPOkPQh4FLgwYg4Engw3TczsxrJrfBH5tV0tyndAvgUsCS1LwFm55XBzMz2lWsfv6QxkrqArcADEfEocFhEbAFIP99WZtkFkjoldRZlvA0zs1rItfBHxO6IaAOmASdJeu8gll0UEe0R0d7S0pJbRjOzoqnJWT0R8QqwGjgDeFHSVID0c2stMpiZWSbPs3paJB2apscCpwNPAncD89Ns84G78spgZmb7ynNY5qnAEkljyN5glkfEPZIeBpZL+hzwS+AzOWYwM7M+civ8EdENHN9P+3bgtLy2a2ZmA/M3d83MCsaF38ysYFz4zcwKxoXfzKxgXPjNzArGhd/MrGBc+M3MCsaF38ysYFz4zcwKxoXfzKxgXPjNzArGhd/MrGBc+M3MCsaF38ysYFz4zcwKxoXfzKxgXPjNzArGhd/MrGBc+M3MCia3wi9puqRVkjZKekLSF1N7h6TnJXWl25l5ZTAzs33ldrF1YBfw1xHxmKRxwDpJD6THro6Iq3LctpmZlZFb4Y+ILcCWNL1D0kbg7Xltz8zMKlOTPn5JrcDxwKOp6SJJ3ZJulDSxzDILJHVK6ty2bVstYpqZFULuhV/SIcD3gYUR8VvgOuBdQBvZfwTf6m+5iFgUEe0R0d7S0pJ3TDOzwsi18EtqIiv6t0TECoCIeDEidkfEG8D1wEl5ZjAzs73leVaPgO8CGyPi70rap5bMNgfYkFcGMzPbV55n9ZwMnAs8LqkrtX0VOEdSGxDAZuDzOWYwM7M+8jyr58eA+nno3ry2aWZm++dv7pqZFYwLv5lZwbjwm5kVjAu/mVnBuPCbmRVMnqdzmtVOd0d9tnvcvtvt2LepavJctxWHj/jNzArGhd/MrGBc+M3MCsaF38ysYAZd+CVNlHRcHmHMzCx/FZ3VI2k18Mk0fxewTdKaiLg4v2hm+Vm9pkrrWVGd9ZjVUqVH/BPSRVQ+DfxDRJwAnJ5fLDMzy0ulhf+ANI7+WcA9OeYxM7OcVVr4Lwd+CDwdEWslvRN4Kr9YZmaWl0q/ubslIvZ8oBsRmyT93UALmJlZY6r0iP/bFbaZmVmDG/CIX9KHgY8ALZJKz+AZD4zJM5iZmeVjf109BwKHpPnGlbT/FpibVygzM8vPgIU/ItYAayQtjohnB7NiSdOB7wH/CXgDWBQR/1vSJOBWoJXsYutnRcTLQ8huZmZDUOmHuwdJWkRWrPcsExGnDrDMLuCvI+IxSeOAdZIeAM4DHoyIKyRdClwKXDKU8GZmNniVFv7bgO8ANwC7K1kgIrYAW9L0DkkbgbcDnwJOSbMtAVbjwm9mVjOVFv5dEXHdUDciqRU4HngUOCy9KRARWyS9rcwyC4AFADNmzBjqps3MrI9KT+dcKel/SpoqaVLvrZIFJR0CfB9YmIZ9qEhELIqI9ohob2lpqXQxMzPbj0qP+Oenn18paQvgnQMtJKmJrOjfEhG9w1m9KGlqOtqfCmwdTGAzMxueigp/RMwc7IolCfgusDEiSr/lezfZG8kV6eddg123mZkNXaXDMv+3/toj4nsDLHYycC7wuKSu1PZVsoK/XNLngF8Cn6k4rZmZDVulXT0nlkw3A6cBj5Gdp9+viPgxoDIPn1bhds3MrMoq7er5Qul9SROAm3JJZGZmuRrqNXd/BxxZzSBmZlYblfbxryQ7iweywdmOAZbnFcrMzPJTaR//VSXTu4BnI6InhzxmZpazirp60mBtT5KN0DkR+EOeoczMLD+VdvWcBfwt2bg6Ar4t6SsRcXuO2azgVq+pdwKz0anSrp6vASdGxFYASS3APwMu/GZmI0ylZ/W8pbfoJ9sHsayZmTWQSo/475P0Q2Bpuj8PuDefSGZmlqf9XXP3CLJhlL8i6dPAH5H18T8M3FKDfGZmVmX76665BtgBEBErIuLiiPgS2dH+NflGMzOzPOyv8LdGRHffxojoJLsMo5mZjTD7K/zNAzw2tppBzMysNvZX+NdK+h99G9OQyuvyiWRmZnna31k9C4E7JH2WNwt9O3AgMCfHXGZmlpMBC39EvAh8RNJHgfem5n+KiH/JPZmZmeWi0vH4VwGrcs5iZvvR0TEy122Nxd++NTMrGBd+M7OCya3wS7pR0lZJG0raOiQ9L6kr3c7Ma/tmZta/PI/4FwNn9NN+dUS0pZvH+zEzq7HcCn9E/Ah4Ka/1m5nZ0NSjj/8iSd2pK2hiuZkkLZDUKalz27ZttcxnZjaq1brwXwe8C2gDtgDfKjdjRCyKiPaIaG9paalRPDOz0a+mhT8iXoyI3RHxBnA9cFItt29mZjUu/JKmltydA2woN6+ZmeWj0itwDZqkpcApwBRJPcD/Ak6R1AYEsBn4fF7bNzOz/uVW+CPinH6av5vX9szMrDL+5q6ZWcG48JuZFYwLv5lZwbjwm5kVjAu/mVnB5HZWj1kRnDK5o27bXr29ftu2kc1H/GZmBePCb2ZWMC78ZmYF48JvZlYwLvxmZgXjs3pGo+6OeicwswbmI34zs4Jx4TczKxgXfjOzgnHhNzMrGBd+M7OCceE3MysYF34zs4LJrfBLulHSVkkbStomSXpA0lPp58S8tm9mZv3L84h/MXBGn7ZLgQcj4kjgwXTfzMxqKLfCHxE/Al7q0/wpYEmaXgLMzmv7ZmbWv1r38R8WEVsA0s+3lZtR0gJJnZI6t23bVrOAZmajXcN+uBsRiyKiPSLaW1pa6h3HzGzUqHXhf1HSVID0c2uNt29mVni1Lvx3A/PT9Hzgrhpv38ys8PI8nXMp8DBwtKQeSZ8DrgD+VNJTwJ+m+2ZmVkO5jccfEeeUeei0vLZpZmb717Af7pqZWT5c+M3MCsaF38ysYFz4zcwKxoXfzKxgXPjNzArGhd/MrGBc+M3MCsaF38ysYFz4zcwKxoXfzKxgchurx4ph9Zp6J7Bq6egYmeu2wfMRv5lZwbjwm5kVjAu/mVnBuPCbmRWMC7+ZWcG48JuZFYwLv5lZwdTlPH5Jm4EdwG5gV0S01yOHmVkR1fMLXB+NiF/XcftmZoXkrh4zs4Kp1xF/APdLCuD/RcSivjNIWgAsAJgxY0aN41VJd0e9E5iZ7aNeR/wnR8QHgI8DF0r6474zRMSiiGiPiPaWlpbaJzQzG6XqUvgj4oX0cytwB3BSPXKYmRVRzQu/pLdKGtc7DfwZsKHWOczMiqoeffyHAXdI6t3+P0bEfXXIYWZWSDUv/BGxCXh/rbdrNtqcMrmj3hEq112l9RzXUaUVFZtP5zQzKxgXfjOzgnHhNzMrGBd+M7OCceE3MyuYeg7SNuJ0dAxu/lMmD2LePxncus3MhspH/GZmBePCb2ZWMC78ZmYF48JvZlYwLvxmZgXjs3oaxOo1+a3bZwyZWSkf8ZuZFYwLv5lZwbjwm5kVjAu/mVnBjP4Pd7s7qraqwQzB0Ejy/ODYrBLV+htcvWLftsEOpTIk3bXYSBk5XHzGR/xmZgXjwm9mVjB1KfySzpD0C0lPS7q0HhnMzIqq5oVf0hjg/wIfB44FzpF0bK1zmJkVVT2O+E8Cno6ITRHxB2AZ8Kk65DAzKyRFRG03KM0FzoiIv0j3zwU+GBEX9ZlvAbAg3T0a+EWOsaYAv85x/dXkrNU3UnKCs+ZlpGQdbM53RERL38Z6nM6pftr2efeJiEXAovzjgKTOiGivxbaGy1mrb6TkBGfNy0jJWq2c9ejq6QGml9yfBrxQhxxmZoVUj8K/FjhS0kxJBwJnA3fXIYeZWSHVvKsnInZJugj4ITAGuDEinqh1jj5q0qVUJc5afSMlJzhrXkZK1qrkrPmHu2ZmVl/+5q6ZWcG48JuZFcyoL/z7Gx5C0lckdaXbBkm7JU0qeXyMpH+VdE+j5pR0qKTbJT0paaOkDzdw1i9JeiK1L5XUXOesEyStlLQ+5Tq/0mUbJauk6ZJWpd/9E5K+2Ig5Sx6vyT413KwNuF8NlHVw+1VEjNob2YfH/w68EzgQWA8cO8D8s4B/6dN2MfCPwD2NmhNYAvxFmj4QOLQRswJvB54Bxqb7y4Hz6pkV+CpwZZpuAV5K8w7qedY561TgA6l9HPBveWUdTs6Sx3Pfp6qRtdH2qwF+/4Per0b7Ef9gh4c4B1jae0fSNOC/ADfkmnIYOSWNB/4Y+C5ARPwhIl5pxKzJAcBYSQcAB5PvdzgqyRrAOEkCDiHbmXZVuGxDZI2ILRHxGEBE7AA2khWDhsoJNd2nhpW1Qfersq8rg9yvRnvhfzvwXMn9HsrsEJIOBs4Avl/SfA3wN8AbOeXrNZyc7wS2Af+Q/n2+QdJbGzFrRDwPXAX8EtgC/CYi7q9z1muBY8h2lMeBL0bEGxUuW03DybqHpFbgeODRBs15DbXZp2B4WRtxv+o361D2q9Fe+CsaHiKZBfwkIl4CkPQJYGtErMsrXIkh5yR7p/8AcF1EHA/8B5Bnf/RwXtOJZEcxM4HDgbdK+vNcUmYqyfoxoCvlaQOuTUd7g3me1TCcrNkKpEPI3mQXRsRv84k59Jw13qdgeK9pI+5X5V7XQe9Xo73wD2Z4iLPZu0viZOCTkjaT/dt1qqSb8wjJ8HL2AD0R0XuEdzvZH2xehpP1dOCZiNgWEa8DK4CP5JIyU0nW84EVkXmarK/03RUu2yhZkdREVvRviYh+LlDYEDlruU8NN2sj7lflsg5+v8rrw4pGuJG9a28ieyfs/cDkPf3MN4Gsv+ytZdZzCvl+uDusnMBDwNFpugP420bMCnwQeIKsD1JkH559oZ5ZgeuAjjR9GPA82QiIFT3PBskq4HvANXnlq0bOPvPkuk9VI2uj7VcD/P4HvV/l+kfSCDfgTLKzHP4d+FpquwC4oGSe84BlA6yjFn+kQ85J9m9fJ9AN3AlMbOCslwNPAhuAm4CD6pmV7F/j+8n6TDcAfz7Qso2YFfgjsm6BbrKugC7gzEbL2Wcdue9TVfj9N9R+tZ+sg9qvPGSDmVnBjPY+fjMz68OF38ysYFz4zcwKxoXfzKxgXPjNzArGhd8qlkbZ7Cq5tQ5zfW2Sziy5/8lajII5GJJWSxrSxa0lLUzDVvTev1fSof3M1yHpy8OIOdhc50k6vIJ5rq1VJqutml960Ua030dEW38PpIGjFH3GjtmPNqAduBcgIu5mdF1/eSFwM/A7gIg4c8C5a+c8svO98/wmsjUwH/HbkElqTeOU/z3wGDBd0nWSOtPY4JeXzHuipJ+mscR/JmkC8A1gXvrvYV7pUaakd0h6UFJ3+jkjtS+W9H/SujZJmjtArutTjvsljU2P7TmClzQlDR/QO0b8VZIeT9v8Qj/r/TNJD0t6TNJtaWwcJJ2WBvJ6XNKNkg6S9FdkX7hZJWlVmm+zpClp+mvKxl7/Z+Dokm28S9J9ktZJekjSu/vJ0SFpSXpemyV9WtI30/bvS8M3IOkySWuVjdG+SJm5ZG+2t6TXfWw/v5txaVOHp/U9JembFbwOV0j6eXr9rqrgT8jqJe9vzvk2em7Abt78ZugdQCvZKIsfKplnUvo5BlgNHEf2FfRNwInpsd5BsM4Dri1Zds99YCUwP03/d+DONL0YuI3soOVYsqFs++ZsJRuuti3dX86b33JdDbSn6SnA5jT9l2Rj3RzQ53msJiuUU4AfkYagAC4BLgOayUZVPCq1f49skDSAzZQMVdB7HziB7NuXB6fX4mngy2meB4Ej0/QH6XN9iNTeAfwYaALeT/YfxcfTY3cAs0ufQ5q+CZjVz2sw0O9mE9nQG83As2RjyZR7HSYBv+DN63gfWu+/V9/K39zVY4OxV1dP6uN/NiIeKZnnLEkLyIrHVLLiHMCWiFgLEGnkyKx3qKwPA59O0zcB3yx57M7IupR+LumwMss/ExFdaXod2ZvBQE4HvhMRu1LGl/o8/qH0XH6Sch8IPEx2tP5MRPxbmm8JcCHZ8MPl/Gfgjoj4HYCku9PPQ8gG17qt5LU5qMw6fhARr0t6nOxN9r7U/njJc/2opL8he4OZRDaey8o+6zma8r+bByPiN+n+z4F3AIeWeR1+C+wEbpD0T0DuV9eyoXPht+H6j94JSTOBL5MdPb4saTHZ0aIY/pDGpcu/VjJd7t2jdJ7dwNg0vYs3uzhLL0+3v4wCHoiIc/ZqlNoGWGYg/W3rLcArUeZzlD5eA4iINyS9Hukwm+w/sAOUXXrv78mO7J+T1MHez7fXQM+772t4AGVeBwBJJwGnkY3KehFwagXPw+rAffxWTePJ3gh+k47EP57anyTrLz4RQNI4ZVcK2kF2qcD+/JSsgAB8lqxroxo2k3W1AJR+PnA/cEHKhUquu5w8Apws6Yj0+MGSjiJ7bq297cC5wJo0Xe75/QiYk/rXx5Fdt6D3aPsZSZ9J25Ck9w/xefYW+V+n/yRKn2tprnK/m3L6fR3SNiZExL1kH2q3DTG31YALv1VNRKwH/pWsS+FG4Cep/Q/APODbktYDD5AVplXAselDxnl9VvdXwPmSusmK6RerFPMq4C8l/ZSsv7rXDWRXMOpOGf9rn+e2jazfe2nK9Ajw7ojYSTZO+m2p2+UN4DtpsUXAD3o/3C1Z12PArWSflXyfbPjfXp8FPpcyPMEQL/cY2WUCryfr+rkTWFvy8GLgO5K6yLqJ+vvdlFtvv68D2RvJPaltDfCloeS22vDonGZmBeMjfjOzgnHhNzMrGBd+M7OCceE3MysYF34zs4Jx4TczKxgXfjOzgvn/bQ/sz72YakYAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.hist(hmm_fms, 10, facecolor='blue', alpha=0.5, label='HMM')\n",
    "plt.hist(cu_fms, 10, facecolor='orange', alpha=0.5, label='CU')\n",
    "plt.legend()\n",
    "plt.xlabel('Fraction nucleotide matches')\n",
    "plt.ylabel('Counts');"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6.656586080520897e-07"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "_, pval = ttest_rel(hmm_fms, cu_fms, alternative = \"greater\")\n",
    "pval"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The HMM is outperforming the codon usage approach by a statistically significant margin."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
